### PR TITLE
[FLINK-6008][blob] collection of BlobServer improvements

### PR DESCRIPTION
**Pull request by NicoK**
This PR is a light-weight version of #3512 that only includes the improvements and can serve as a base for FLIP-19. It improves the following things around the `BlobServer`/`BlobCache`:

* replace config options in `config.md` with non-deprecated ones, e.g. `high-availability.cluster-id` and `high-availability.storageDir`
* do not fail the `BlobServer` when a delete operation fails
* add more unit tests
* general code style and docs improvements, like using `Preconditions.checkArgument`

### CODE REVIEW

**Comment by NicoK** [comment]
please note that there are four unrelated test failures:
* 2x Kafka010ITCase
* 2x YARNSessionCapacitySchedulerITCase

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServerConnection.java [hunk 1]
```diff
@@ -509,21 +509,7 @@ private void delete(InputStream inputStream, OutputStream outputStream, byte[] b
...
 				BlobKey key = BlobKey.readFromInputStream(inputStream);
-				File blobFile = blobServer.getStorageLocation(key);
-
-				writeLock.lock();
-
-				try {
-					// we should make the local and remote file deletion atomic, otherwise we might risk not
-					// removing the remote file in case of a concurrent put operation
-					if (blobFile.exists() && !blobFile.delete()) {
-						throw new IOException("Cannot delete BLOB file " + blobFile.getAbsolutePath());
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServerConnection.java]: Don't you want to add a test coverage for not throwing if delete fails?

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java [hunk 2]
```diff
@@ -227,7 +227,7 @@ static File getStorageLocation(File storageDir, JobID jobID, String key) {
 	private static File getJobDirectory(File storageDir, JobID jobID) {
 		final File jobDirectory = new File(storageDir, JOB_DIR_PREFIX + jobID.toString());
 
-		if (!jobDirectory.exists() && !jobDirectory.mkdirs()) {
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java]: Why did you change the order of those operations? Is that fix for something? If so could you add explanation to the commit message what's going on here?

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerPutTest.java [hunk 3]
```diff
@@ -59,6 +60,110 @@
...
+		private final BlobKey key;
+
+		public ContentAddressableGetStorageLocation(BlobServer server, BlobKey key) {
+			this.server = server;
+			this.key = key;
+		}
+
+		@Override
+		public void go() throws Exception {
+			server.getStorageLocation(key);
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerPutTest.java]: Shouldn't you call this at least couple/couple of dozens/couple of hundred times? Otherwise won't this complete before next thread starts up?

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java [hunk 4]
```diff
@@ -411,12 +411,11 @@ public void delete(BlobKey key) throws IOException {
...
 		try {
-			if (localFile.exists()) {
-				if (!localFile.delete()) {
-					LOG.warn("Failed to delete locally BLOB " + key + " at " + localFile.getAbsolutePath());
-				}
+			if (!localFile.delete() && localFile.exists()) {
+				LOG.warn("Failed to delete locally BLOB " + key + " at " + localFile.getAbsolutePath());
 			}
 
+
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java]: remove extra line

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java [hunk 5]
```diff
@@ -593,9 +594,7 @@ private void sendPutHeader(OutputStream outputStream, JobID jobID, String key) t
 	 *         the BLOB server or if the BLOB server cannot delete the file
 	 */
 	public void delete(BlobKey key) throws IOException {
-		if (key == null) {
-			throw new IllegalArgumentException("BLOB key must not be null");
-		}
+		checkArgument(key != null, "BLOB key must not be null.");
```
Comment (NicoK) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java]: unfortunately, `checkNotNull` throws a `NullPointerException` instead and I did not want to change that here

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java (hunk 2 above)
Comment (NicoK) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java]: the new way is thread-safe - imagine a directory being concurrently created after another thread has failed the `exists()` part - not sure we actually fix a bug nowadays since the use may be guarded...so this is probably more of a "just-in-case" thing or if usage patterns change

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerPutTest.java (hunk 3 above)
Comment (NicoK) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerPutTest.java]: Actually, the job directory is only created once so it doesn't help adding more calls. What could help make this happen more often, is to add more threads to `BlobServerPutTest#testServerContentAddressableGetStorageLocationConcurrent()` otherwise the failure only happens once in a while if the implementation is not thread-safe

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServerConnection.java (hunk 1 above)
Comment (NicoK) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServerConnection.java]: the (included) changes in `BlobServerDeleteTest` should cover this

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java (hunk 5 above)
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java]: Yes, but it would be cleaner and less surprising for future users. However I have no strong feelings about it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServerConnection.java (hunk 1 above)
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServerConnection.java]: Ok, sorry, I didn't find it first time I was looking for it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java (hunk 2 above)
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java]: Ok, please add comment or commit message info about this.

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerPutTest.java (hunk 3 above)
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerPutTest.java]: So maybe you could call in a loop multiple times `gets` and `deletes` (if there is such operation) interleaved? Otherwise I don't see a real value of those tests and I would prefer to drop them (so that we don't have to maintain tests that do not check for anything).

Comment (NicoK) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerPutTest.java]: Unfortunately, concurrency with `delete` operations does not work either if not guarded - the directory may not exist anymore between `jobDirectory.mkdirs()` and `jobDirectory.exists()`. I was able to reproduce the error with the existing test though - if you want to try it, just change the order of these two commands back - the test will not hit every time, but some times.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServerConnection.java [hunk 6]
```diff
@@ -540,7 +526,7 @@ else if (type == NAME_ADDRESSABLE) {
 					// we should make the local and remote file deletion atomic, otherwise we might risk not
 					// removing the remote file in case of a concurrent put operation
 					if (blobFile.exists() && !blobFile.delete()) {
```
Comment (NicoK) [inline flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServerConnection.java]: you're right, this could be improved as well - the next PR in this series is removing this code though

File: flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerDeleteTest.java [hunk 7]
```diff
@@ -92,6 +99,69 @@ public void testDeleteSingle() {
...
+				fail("BLOB should have been deleted");
+			}
+			catch (IOException e) {
+				// expected
+			}
+		}
+		catch (Exception e) {
+			e.printStackTrace();
+			fail(e.getMessage());
+		}
```
Comment (NicoK) [inline flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerDeleteTest.java]: Absolutely right - I was told this was an issue some versions ago with JUnit or so... was just copying the code from the other tests in this class. Let me create a cleanup-PR on top of the newest PR in this series, i.e. #4238

**Comment by zentol** [comment]
merging,
